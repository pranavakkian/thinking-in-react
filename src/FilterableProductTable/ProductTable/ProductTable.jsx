import PropTypes from 'prop-types';
import React, { Fragment, memo } from 'react';
import ProductCategoryRow from './ProductCategoryRow/ProductCategoryRow';
import ProductRow from './ProductRow/ProductRow';

const ProductTable = (props) => {
  const { data } = props;
  const headingSet = new Set();
  data.map((item) => headingSet.add(item.category));
  const headings = [];
  headingSet.forEach((item) => headings.push(item));

  return (
    <table>
      <thead>
        <tr>
          <th id="nameHeader">Name</th>
          <th id="priceHeader">Price</th>
        </tr>
      </thead>
      <tbody>
        {headings.map((heading) => (
          <Fragment key={heading}>
            <tr>
              <ProductCategoryRow category={heading} />
            </tr>
            {data
              .filter((item) => item.category === heading)
              .map((value) => (
                <tr key={value.name}>
                  <ProductRow data={value} />
                </tr>
              ))}
          </Fragment>
        ))}
      </tbody>
    </table>
  );
};

ProductTable.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      category: PropTypes.string.isRequired,
      price: PropTypes.string.isRequired,
      stocked: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default memo(ProductTable);
