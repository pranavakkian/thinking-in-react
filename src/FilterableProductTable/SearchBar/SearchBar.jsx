import React, { memo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  handleFilterTextChange,
  handleIsStockOnlyChange,
} from '../Redux/SearchBarReducer/Actions/Actions';

const SearchBar = () => {
  const dispatch = useDispatch();

  return (
    <>
      <input
        type="text"
        value={useSelector((state) => state.SearchBarReducer.filterText)}
        onChange={(event) => dispatch(handleFilterTextChange(event.target.value))}
        name="filterText"
        id="filterText"
        placeholder="Search product..."
      />
      <br />
      <input
        checked={useSelector((state) => state.SearchBarReducer.isStockOnly)}
        onChange={(event) => dispatch(handleIsStockOnlyChange(event.target.checked))}
        type="checkbox"
        name="isStockOnly"
        id="isStockOnly"
      />
      <span id="isStockOnlyLabel">Only show product in stock</span>
    </>
  );
};

export default memo(SearchBar);
