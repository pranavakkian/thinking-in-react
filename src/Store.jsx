import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import RootReducer from './RootReducer';

const composeEnhancers = composeWithDevTools({ trace: true });
const Store = createStore(RootReducer, composeEnhancers(applyMiddleware()));

export default Store;
