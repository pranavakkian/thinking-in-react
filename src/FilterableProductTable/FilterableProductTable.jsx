import axios from 'axios';
import React, { memo, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ProductTable from './ProductTable/ProductTable';
import { handleCleanUpSearchBarReducer } from './Redux/SearchBarReducer/Actions/Actions';
import SearchBar from './SearchBar/SearchBar';

const FilterableProductTable = () => {
  const [data, setData] = useState();
  const [filteredData, setFilteredData] = useState();
  const dispatch = useDispatch();
  const filterText = useSelector((state) => state.SearchBarReducer.filterText);
  const isStockOnly = useSelector(
    (state) => state.SearchBarReducer.isStockOnly,
  );

  useEffect(() => {
    axios.get('').then((response) => {
      setData(response.data);
      setFilteredData(response.data);
    });
    return () => {
      dispatch(handleCleanUpSearchBarReducer());
    };
    // eslint-disable-next-line
  }, []);

  useEffect(() => {
    if (filterText !== '' && isStockOnly) {
      setFilteredData(
        data.filter(
          (item) => item.name.toLocaleLowerCase().includes(filterText)
            /* istanbul ignore next */
            && item.stocked,
        ),
      );
    } else if (filterText !== '' && !isStockOnly) {
      setFilteredData(
        data.filter((item) => item.name.toLowerCase().includes(filterText)),
      );
    } else if (filterText === '' && isStockOnly) {
      setFilteredData(data.filter((item) => item.stocked));
    } else {
      setFilteredData(data);
    }
    // eslint-disable-next-line
  }, [filterText, isStockOnly]);

  return (
    <>
      <SearchBar />
      {filteredData && <ProductTable data={filteredData} />}
    </>
  );
};

export default memo(FilterableProductTable);
