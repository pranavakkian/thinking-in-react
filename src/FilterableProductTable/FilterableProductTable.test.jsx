/* eslint-disable no-undef */
import axios from 'axios';
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import Store from '../Store';
import FilterableProductTable from './FilterableProductTable';
import Mockdata from './MockData/MockData';
import ProductTable from './ProductTable/ProductTable';
import SearchBar from './SearchBar/SearchBar';

const data = Mockdata;
const store = Store;
jest.mock('axios');
axios.get.mockImplementation(() => Promise.resolve(data));
const wrapper = mount(
  <Provider store={store}>
    <FilterableProductTable />
  </Provider>,
);

describe('FilterableProductTable', () => {
  it('FilterableProductTable rendered successfully!', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Child components rendered successfully!', () => {
    expect(wrapper.contains(<SearchBar />)).toBeTruthy();
    expect(wrapper.contains(<ProductTable data={data} />)).toBeTruthy();
  });

  it('filterText not empty and isStockOnly=true', () => {
    wrapper
      .find('#filterText')
      .simulate('change', { target: { value: 'Football' } });
    wrapper
      .find('#isStockOnly')
      .simulate('change', { target: { checked: true } });
    expect(store.getState().SearchBarReducer).toEqual({
      filterText: 'Football',
      isStockOnly: true,
    });
    expect(
      wrapper.contains(
        <ProductTable
          data={data.filter(
            (item) => item.name.toLocaleLowerCase().includes('Football') && item.stocked,
          )}
        />,
      ),
    ).toBeTruthy();

    wrapper
      .find('#filterText')
      .simulate('change', { target: { value: 'Basketball' } });
    wrapper
      .find('#isStockOnly')
      .simulate('change', { target: { checked: true } });
    expect(store.getState().SearchBarReducer).toEqual({
      filterText: 'Basketball',
      isStockOnly: true,
    });
    expect(
      wrapper.contains(
        <ProductTable
          data={data.filter(
            (item) => item.name.toLocaleLowerCase().includes('Basketball')
              && item.stocked,
          )}
        />,
      ),
    ).toBeTruthy();
  });

  it('filterText not empty and isStockOnly=false', () => {
    wrapper
      .find('#filterText')
      .simulate('change', { target: { value: 'Football' } });
    wrapper
      .find('#isStockOnly')
      .simulate('change', { target: { checked: false } });
    expect(store.getState().SearchBarReducer).toEqual({
      filterText: 'Football',
      isStockOnly: false,
    });
    expect(
      wrapper.contains(
        <ProductTable
          data={data.filter((item) => item.name.toLowerCase().includes('Football'))}
        />,
      ),
    ).toBeTruthy();
  });

  it('filterText empty and isStockOnly=true', () => {
    wrapper.find('#filterText').simulate('change', { target: { value: '' } });
    wrapper
      .find('#isStockOnly')
      .simulate('change', { target: { checked: true } });
    expect(store.getState().SearchBarReducer).toEqual({
      filterText: '',
      isStockOnly: true,
    });
    expect(
      wrapper.contains(
        <ProductTable data={data.filter((item) => item.stocked)} />,
      ),
    ).toBeTruthy();
  });

  it('FilterableProductTable un-mounted successfully!', () => {
    wrapper.unmount();
    expect(store.getState().SearchBarReducer).toEqual({
      filterText: '',
      isStockOnly: false,
    });
  });
});
