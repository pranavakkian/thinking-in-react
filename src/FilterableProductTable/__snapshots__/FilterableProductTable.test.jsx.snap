// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableProduct FilterableProductTable rendered successfully! 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Memo(FilterableProductTable)>
    <Memo(SearchBar)>
      <input
        id="filterText"
        name="filterText"
        onChange={[Function]}
        placeholder="Search product..."
        type="text"
        value=""
      />
      <br />
      <input
        checked={false}
        id="isStockOnly"
        name="isStockOnly"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        id="isStockOnlyLabel"
      >
        Only show product in stock
      </span>
    </Memo(SearchBar)>
    <Memo(ProductTable)
      data={
        Array [
          Object {
            "category": "Sporting Goods",
            "name": "Football",
            "price": "$49.99",
            "stocked": true,
          },
          Object {
            "category": "Sporting Goods",
            "name": "Baseball",
            "price": "$9.99",
            "stocked": true,
          },
          Object {
            "category": "Sporting Goods",
            "name": "Basketball",
            "price": "$29.99",
            "stocked": false,
          },
          Object {
            "category": "Electronics",
            "name": "iPod Touch",
            "price": "$99.99",
            "stocked": true,
          },
          Object {
            "category": "Electronics",
            "name": "iPhone 5",
            "price": "$399.99",
            "stocked": false,
          },
          Object {
            "category": "Electronics",
            "name": "Nexus 7",
            "price": "$199.99",
            "stocked": true,
          },
        ]
      }
    >
      <table>
        <thead>
          <tr>
            <th
              id="nameHeader"
            >
              Name
            </th>
            <th
              id="priceHeader"
            >
              Price
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <Memo(ProductCategoryRow)
              category="Sporting Goods"
            >
              <th
                colSpan="2"
                id="categoryHeader"
              >
                Sporting Goods
              </th>
            </Memo(ProductCategoryRow)>
          </tr>
          <tr
            key="Football"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Sporting Goods",
                  "name": "Football",
                  "price": "$49.99",
                  "stocked": true,
                }
              }
            >
              <td
                id="dataName"
              >
                Football
              </td>
              <td
                id="dataPrice"
              >
                $49.99
              </td>
            </Memo(ProductRow)>
          </tr>
          <tr
            key="Baseball"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Sporting Goods",
                  "name": "Baseball",
                  "price": "$9.99",
                  "stocked": true,
                }
              }
            >
              <td
                id="dataName"
              >
                Baseball
              </td>
              <td
                id="dataPrice"
              >
                $9.99
              </td>
            </Memo(ProductRow)>
          </tr>
          <tr
            key="Basketball"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Sporting Goods",
                  "name": "Basketball",
                  "price": "$29.99",
                  "stocked": false,
                }
              }
            >
              <td
                id="dataName"
              >
                Basketball
              </td>
              <td
                id="dataPrice"
              >
                $29.99
              </td>
            </Memo(ProductRow)>
          </tr>
          <tr>
            <Memo(ProductCategoryRow)
              category="Electronics"
            >
              <th
                colSpan="2"
                id="categoryHeader"
              >
                Electronics
              </th>
            </Memo(ProductCategoryRow)>
          </tr>
          <tr
            key="iPod Touch"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Electronics",
                  "name": "iPod Touch",
                  "price": "$99.99",
                  "stocked": true,
                }
              }
            >
              <td
                id="dataName"
              >
                iPod Touch
              </td>
              <td
                id="dataPrice"
              >
                $99.99
              </td>
            </Memo(ProductRow)>
          </tr>
          <tr
            key="iPhone 5"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Electronics",
                  "name": "iPhone 5",
                  "price": "$399.99",
                  "stocked": false,
                }
              }
            >
              <td
                id="dataName"
              >
                iPhone 5
              </td>
              <td
                id="dataPrice"
              >
                $399.99
              </td>
            </Memo(ProductRow)>
          </tr>
          <tr
            key="Nexus 7"
          >
            <Memo(ProductRow)
              data={
                Object {
                  "category": "Electronics",
                  "name": "Nexus 7",
                  "price": "$199.99",
                  "stocked": true,
                }
              }
            >
              <td
                id="dataName"
              >
                Nexus 7
              </td>
              <td
                id="dataPrice"
              >
                $199.99
              </td>
            </Memo(ProductRow)>
          </tr>
        </tbody>
      </table>
    </Memo(ProductTable)>
  </Memo(FilterableProductTable)>
</Provider>
`;
